### PR TITLE
[ticket/15257] Provide extension not enableable messages

### PR DESCRIPTION
**Pull request by javiexin**
Extensions may define the reasons why they cannot be enabled,
so that users are better informed and ready to fix this.
Multiple messages are possible at the same time.
Keeps 100% BC, even with the current "workarounds".

PHPBB3-15257

Checklist:

- [x] Correct branch: master for new features; 3.2.x, 3.1.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master / 3.2.x](https://area51.phpbb.com/docs/master/coding-guidelines.html), [3.1.x](https://area51.phpbb.com/docs/31x/coding-guidelines.html)
- [x] Commit follows commit message [format](https://wiki.phpbb.com/Git#Commit_Messages)

Tracker ticket:

https://tracker.phpbb.com/browse/PHPBB3-15257

### CODE REVIEW

File: phpBB/phpbb/extension/extension_interface.php [hunk 1]
```diff
@@ -27,6 +27,13 @@
 	public function is_enableable();
 
 	/**
+	* Indicate why the extension can not be enabled.
+	*
+	* @return string|array with actual processed language string(s) (not language keys), or empty
+	*/
+	public function get_not_enableable_msg();
```
Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: I agree with you: the name is terribly ugly... but yours is not much better 😄.  I prefer `not_enableable_msg` if you don't mind, and yes, I will change it.

Regarding the CLI, I was under the impression that this was not considered relevant there.  In fact, the `is_enableable` method is not called (directly) at all in the CLI, it calls `enable` directly (also without going step by step).  If you think this is benefitial, I am happy to add support for it in the CLI, no problem.

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: Done.  Please, review specially the CLI bit, as I am not so experienced there.

Comment (rxu) [inline phpBB/phpbb/extension/extension_interface.php]: Honestly, I'd go with just ```get_message()``` or similar simple name.

**Comment by VSEphpbb** [comment]
You should add some unit tests for this

File: phpBB/phpbb/extension/extension_interface.php (hunk 1 above)
Comment (3D-I) [inline phpBB/phpbb/extension/extension_interface.php]: What about _not_enableable_reason_ ¿

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: I agree, name changed again

**Comment by javiexin** [comment]
@VSEphpbb please, take a look at the added tests, and tell me if ok, or some more are needed (and what do you recommend testing on top of this).

File: tests/functional/extension_acp_test.php [hunk 2]
```diff
@@ -168,6 +168,10 @@ public function test_enable_pre()
 		// Correctly submit the enable form
 		$crawler = self::request('GET', 'adm/index.php?i=acp_extensions&mode=main&action=enable_pre&ext_name=vendor3%2Ffoo&sid=' . $this->sid);
 		$this->assertContainsLang('EXTENSION_NOT_ENABLEABLE', $crawler->filter('.errorbox')->text());
+
+		// Custom error messages returned by not enableable extension
+		$crawler = self::request('GET', 'adm/index.php?i=acp_extensions&mode=main&action=enable_pre&ext_name=vendor4%2Ffoo&sid=' . $this->sid);
+		$this->assertContains('Reason 2', $crawler->filter('.errorbox')->text());
```
Comment (VSEphpbb) [inline tests/functional/extension_acp_test.php]: should also check for `Reason 1`

**Comment by VSEphpbb** [comment]
Tests seem fine to me now.

File: phpBB/includes/acp/acp_extensions.php [hunk 3]
```diff
@@ -197,7 +199,9 @@ function main()
 				$extension = $this->ext_manager->get_extension($ext_name);
 				if (!$extension->is_enableable())
 				{
-					trigger_error($this->user->lang['EXTENSION_NOT_ENABLEABLE'] . adm_back_link($this->u_action), E_USER_WARNING);
+					$message = $extension->not_enableable_reason() ?: $this->user->lang['EXTENSION_NOT_ENABLEABLE'];
+					$message = is_array($message) ? implode('<br/>', $message) : $message;
```
Comment (Nicofuma) [inline phpBB/includes/acp/acp_extensions.php]: $message cannot be an array here

File: phpBB/includes/acp/acp_extensions.php [hunk 4]
```diff
@@ -197,7 +199,9 @@ function main()
 				$extension = $this->ext_manager->get_extension($ext_name);
 				if (!$extension->is_enableable())
 				{
-					trigger_error($this->user->lang['EXTENSION_NOT_ENABLEABLE'] . adm_back_link($this->u_action), E_USER_WARNING);
+					$message = $extension->not_enableable_reason() ?: $this->user->lang['EXTENSION_NOT_ENABLEABLE'];
```
Comment (Nicofuma) [inline phpBB/includes/acp/acp_extensions.php]: $this->user->lang()

File: phpBB/includes/acp/acp_extensions.php (hunk 3 above)
Comment (Nicofuma) [inline phpBB/includes/acp/acp_extensions.php]: you should call $this->user->lang() for every entry in the array

File: phpBB/phpbb/extension/extension_interface.php [hunk 5]
```diff
@@ -27,6 +27,13 @@
 	public function is_enableable();
 
 	/**
+	* Indicate why the extension can not be enabled.
+	*
+	* @return string|array with actual processed language string(s) (not language keys), or empty
+	*/
+	public function not_enableable_reason();
```
Comment (Nicofuma) [inline phpBB/phpbb/extension/extension_interface.php]: I think it's better to change the return of is_enableable() to void|string|array

File: phpBB/includes/acp/acp_extensions.php (hunk 4 above)
Comment (javiexin) [inline phpBB/includes/acp/acp_extensions.php]: I just kept the original way it was.  Will change it, no problem.

File: phpBB/includes/acp/acp_extensions.php (hunk 3 above)
Comment (javiexin) [inline phpBB/includes/acp/acp_extensions.php]: Again, I used what was already in the file.  But I will change it, no problem.

File: phpBB/phpbb/extension/extension_interface.php (hunk 5 above)
Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: That would be a tremendous backward compatibility break: a lot of the currently approved extensions would stop working, and a lot of extensions in development, or even abandoned, would fail for this single reason.
It would have been better that way at the very beginning, but not now, in my view.  The intention here was to keep the current interface intact, and that all existing extensions would continue working without any change.  And if some extension wants to change, then it is easy to do it this way, while keeping 100% BC.

Comment (VSEphpbb) [inline phpBB/phpbb/extension/extension_interface.php]: Yeah is_enableable should only return a Boolean.

Comment (Nicofuma) [inline phpBB/phpbb/extension/extension_interface.php]: having another method for the reason is a bad design as you will have to implement the logic twice.
Another solution might to to deprecate is_enableable() and add another method which can return a string or an array of string. Then we can deal with both form to maintain BC (as we would have done if we have changed the return type of the existing method)

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: Don't like it too much, but if you want, we might:

1. Deprecate is_enableable
2. Make not_enableable_reason return either string/array like now, or boolean; any "false" return (false, '', array()) would mean the extension is enableable; any "true" value would mean not enableable
3. Have a default implementation of not_enableable_reason as !is_enableable
4. Change the core to use the new function with reverse logic.

Comment (Nicofuma) [inline phpBB/phpbb/extension/extension_interface.php]: That's exactly what I proposed. But you need a better name than `not_enableable_reason`

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: Open to suggestions. Maybe just `not_enableable`.

Don't agree with the return format: what would be the meaning of [] or ''?
All truish values should mean the same, and all falsish values as well.

Comment (Nicofuma) [inline phpBB/phpbb/extension/extension_interface.php]: No need for complex parsing, that's what we are already supporting :
 - $language->lang_array('string', array)
 - $language->lang('string')

For the language string not loaded because the extension is not available I see the issue. But we could add in the core a bunch of messages corresponding to the most frequent issue (wrong version of php, wrong version of phpbb, missing php ext, missing dependent ext etc...)

- `''` would just be an empty message, that's not our problem as it won't cause any issue.
- `[]` should use the same default than null

About the truish / falsish values I would ignore that. PHP is a complex language with nuance. The weak typing is a feature but we are not forced to use it everywhere and for everything. It will just make things more complex.

A third solution could be to support and handle exceptions: if an exception cannot be enabled, it just throw an exception in a pre_enable() method per example.

Ref: https://github.com/phpbb/phpbb/blob/master/phpBB/phpbb/language/language.php

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: All of the language methods that you mention take a SINGLE entry and process it.
What we would need is a method that would take either a string, an asoc-array (key->array of params) or an array of keys or array of key+params...  This for me is kind of complex parsing, that of course would use the above services, but needs some significant code.

I don't see that much of an issue with extension messages: the requirement is that the appropriate language file is loaded within the "unnamed new method".  And given that that must be done there (the core cannot have all the possible reasons for an extension not being enableable), for the extension itself would not be complex to process the language strings, using the methods you mention, and without the need of more complex processing in the core.

I understand that we can do what we need with values and types, but I honestly believe that things are much easier to understand and comprehend by the general user when "make sense".  So, having "true" being OK, and having 'some string' being KO does not make much sense to me...

About exceptions, although I think this would be a good solution, it has a significant limitation: single message.

Comment (Nicofuma) [inline phpBB/phpbb/extension/extension_interface.php]:
```
$result = $ext->get_errrors();

if ($result === true) {
     // continue processing
} else {
    $errs = [];
    if (is_string($result) || is_array($result)) {
        foreach ((array) $result as $message) {
             if (is_array($message)) {
                 $errs[] = $language->lang_array($message[0]; $message[1]);
             } else {
                 $errs[] = $language->lang($message);
             }
        }
    } else {
        $errs[] = $language->lang('GENERIC_ERROR_MESSAGE');
    }
}
```

It is a bit complex because it has to deal with a lot of different cases but you can simplify it if you force the most complex form for the errors and assume you will never have anything else (may be a bad idea) :

```
$result = $ext->get_errrors();

if ($result === true) {
     // continue processing
} else {
    $errs = [];
        foreach ((array) $result as $message) {
             $errs[] = $language->lang_array($message[0]; $message[1]);
        }
}
```

For your other concern, a string does not mean true, not in the real world and it does not mean false either. It is matter of convention and having true for a success or a list of errors in case of error is a common convention in software design.

But as I also said, it's not really the best convention. This kind of error could be solve another way:
 - pre_enable() / pre_disable() / pre_purge() methods
 - these methods are not expected to return anything
 - these methods do nothing by default
 - these methods can throw an exception to stop the installer from going forward 
 - the exception can implement the phpbb exception_interface, and in that case the language var will be handled
 - if we really want to support multiple error message we can add a new class of exception to deal with it

Comment (rxu) [inline phpBB/phpbb/extension/extension_interface.php]: It seems for me that both points can be satisfied as well as BC retained: if we inject new class property (say, ```$errs = []```), and then just assign it a value(s) inside ```is_enableable()``` in case of ```false```. The ```is_enableable()``` will still return bool, while in case of ```false``` we could get error mesagge with some class method like ```get_errors()``` returning the ```$errs``` propetry value.

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: I clearly prefer rxu's proposal.  Much cleaner, and better BC (perfect, I should say).
But just rename `not_enableable_reason` to `get_errors` and voila!  That's it.
The fact that $errs is in the class is an implementation detail, that might be the norm, but should not be imposed or required.

Comment (rxu) [inline phpBB/phpbb/extension/extension_interface.php]: @javiexin with theupdated  proposal by @Nicofuma it's actually similar way but more elegant 😉

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: Just for the processing of the language strings, ok.  Otherwise, identical to my proposal.
But again, for that, I would add a method to the language object that would take the "complex" structure and get the clean array of translated messages out (using code similar to the one proposed by Nicofuma).  But, in my view, this is not part of this PR.

Comment (Nicofuma) [inline phpBB/phpbb/extension/extension_interface.php]: A such method does not belong in the language manager. There is too few usage and to few logic. Actually there is nothing inside, just a loop.

About the solution with 2 methods, as I said many time, I don't like and I won't validate it myself.

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: Ok, single method proposal.

One single method, called `not_enableable`.  All current uses of `is_enableable` should simply be negated, and that would be 100% same behaviour as now.

If the method returns `true` the extension cannot be enabled, without any specified reason, so the default error string should be used.  If the method returns a (non-empty) string, or a (non-empty) array of strings, then this should be interpreted as PROCESSED language strings, and transmitted as such to the end user.  If the method returns false, an empty string or an empty array, the extension may be enabled normally.

The `base` implementation would implement `not_enableable` as `!is_enableable`, therefore making it 100% BC.  For extensions that have implemented a "workaround" for this limitation, they will need to rewrite, or continue using the "workaround".

To help extension developers with the language processing of the returned string/array, a helper function will be included in the `base` ext object, so that the extension developer may simply do `return $this->helper($array_of_key)`, being the array of keys as complex as described above, with all possible variations.

If this proposal is accepted, I will gladly code it and resubmit, but I won't do anything until the concept is approved.  If this proposal is not accepted, please, let me know, so that I can close this PR.

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: @Nicofuma: should I implement this, or still not ok on your side?

Comment (javiexin) [inline phpBB/phpbb/extension/extension_interface.php]: @Nicofuma if this is not going to progress, better to close it.  Let me know: close it, or implement proposed alternative?

**Comment by 3D-I** [comment]
Guys, Javier is not anymore with us, life's sad sometimes. 
Please take care of his contributions here.

Thanks.
https://www.phpbb.com/community/viewtopic.php?p=14938226#p14938226
